### PR TITLE
Add dask settings to preferences

### PR DESCRIPTION
**Pull request by ppwadhwa**
# Description

This PR exposes the ability to turn dask caching off/on in the preferences menu.  If dask is enabled, then you can also set the cache size.  

Adding dask settings to preferences was mentioned in the list here: #2172.  

TO DO:  write tests

Screen Shot 2022-04-27 at 3 10 14 PM

### CODE REVIEW

**Review by sofroniewn — APPROVED** [review]
LGTM! Maybe @jni and @tlambert03 want to take a look too?

File: napari/utils/_dask_utils.py
Comment (jni) [inline napari/utils/_dask_utils.py]: Is this a new black version, or some kind of error? Maybe we can just revert all changes to this file?

File: napari/_qt/widgets/qt_dask_settings.py
Comment (jni) [inline napari/_qt/widgets/qt_dask_settings.py]: This class seems like **a lot** of code for what it's doing... My feeling is that we could/should implement this as a magicgui widget with a float value. That might not be feasible within this one PR, but basically I'd like to have a clear road towards simpler, even declarative classes for preferences. If every bool + float preference engenders this much code, preferences become a maintenance nightmare that I'd prefer (heh 😜) not to deal with...

File: napari/_qt/widgets/qt_dask_settings.py
Comment (jni) [inline napari/_qt/widgets/qt_dask_settings.py]:
```suggestion
        Bool value indicating if the dask cache is enabled.
```

File: napari/_qt/widgets/qt_dask_settings.py
Comment (jni) [inline napari/_qt/widgets/qt_dask_settings.py]: One example of simplification: perhaps we get rid of inc (value can only be set by typing), and we leave the range checking to the base model rather than the UI? e.g. I can type "5000000" in the box, but when Qt tries to update the model, it hits the max, and that updates the value to the max. (Perhaps raising a warning, though I can live without it.)

File: napari/_qt/qt_viewer.py
Comment (jni) [inline napari/_qt/qt_viewer.py]: I don't like that this function takes no parameters, but rather grabs a global value when triggered. Could this be engineered so that the event contains the specific values, and the connected function takes those values as parameters? If I remember correctly, event values are now unpacked for callbacks? ie the dask configuration dictionary is either passed directly as an argument to the callback, or it's passed in `event.value`, I can't remember. (It's easy to check by using `.connect(print)`.)

Then this function could be something like:

```python
enabled = setting['enabled']
size = setting['cache']
resize_dask_cache(int(enabled) * size * 1e6)
```

**Review by jni — DISMISSED** [review]
Thanks @ppwadhwa! Functionally this is great, but to be honest I find the implementation is too complex for what we want to achieve. It would be great if we could just make a minimal widget here, with just a write-in field, so we don't have to deal with increments, mins, maxes, etc.

Have we made EventedModel fields that are floats in a constrained range? I have a vague memory of this but maybe it was only discussed, not implemented.

Assuming more people are ok with the amount of code here I'm happy to have my review dismissed, but wanted to pause and think about simplification options here.

**Comment by ppwadhwa** [comment]
thanks, @jni, for your comments.  I will look at how to make this more simple based on your feedback.  It was very helpful.  stay tuned!

File: napari/utils/_dask_utils.py
Comment (ppwadhwa) [reply napari/utils/_dask_utils.py]: I had changed this at one point, it created a new version during pre-commit.  I'll change it back.

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (goanpeca) [inline napari/_qt/widgets/magic_gui_dask_settings.py]: Could this code 

```python
class DaskSettings(BaseMagicSetting):
    """Class for use in json schema widget builder for dask settings."""

    def __init__(self):

        self.MAGIC_GUI = dask_settings
        super().__init__()
```

Become

```python
class DaskSettings(BaseMagicSetting):
    """Class for use in json schema widget builder for dask settings."""

    MAGIC_GUI = dask_settings
```

Comment (ppwadhwa) [reply napari/_qt/widgets/magic_gui_dask_settings.py]: @goanpeca I tried your code, and it gives me a `RecursionError`.  The only way I can see to avoid the error is to do what I have above.

File: napari/_qt/qt_viewer.py
Comment (ppwadhwa) [reply napari/_qt/qt_viewer.py]: I made a change here in order to use the event.value in the callback.  I tried to pass in a dictionary only by doing 
```
settings.application.events.dask.connect(lambda x: self._update_dask_settings(x.value))
```
but the tests were failing because something wasn't being cleaned up properly.  So, this function now accepts either an event that must have a dictionary in event.value or it will accept a dictionary.   I'm not sure if there is a better way?

File: napari/_qt/widgets/qt_dask_settings.py
Comment (kephale) [reply napari/_qt/widgets/qt_dask_settings.py]: This point was addressed by @ppwadhwa doing a refactor to make: 

- https://github.com/napari/napari/blob/dbd7b9d32ebd44b15d83dcde0609ae2a8a4c5d4a/napari/_qt/widgets/magic_gui_base_widget.py
- https://github.com/napari/napari/blob/dbd7b9d32ebd44b15d83dcde0609ae2a8a4c5d4a/napari/_qt/widgets/magic_gui_dask_settings.py

File: napari/_qt/widgets/qt_dask_settings.py
Comment (kephale) [reply napari/_qt/widgets/qt_dask_settings.py]: I believe that this is resolved by the widget refactoring.

**Comment by kephale** [comment]
I was able to resolve the conflicts quite easily: https://github.com/kephale/napari/tree/dask-config but cannot submit a PR to https://github.com/ppwadhwa/napari for some reason. 

It looks like @ppwadhwa addressed all of the comments in the review, and as a result added what looks like some nice `magicgui` functionality for adding new settings widgets. 

With the current async slicing work (https://github.com/napari/napari/blob/main/docs/naps/4-async-slicing.md) we're starting to need to mess around Dask caching more, so it would be great to merge this PR.

Is there anything else that folks want done before this can be merged?

**Comment by jni** [comment]
WOW, my sincere apologies to @ppwadhwa for neglecting this for so long! It must have come at a busy period and then gotten buried in the pile, I'm sorry! I've only just had a brief glance at it but it looks so amazing now! Please don't hesitate to pester me a few times (especially in `#reviews` on Zulip) if I drop the ball so thoroughly again in the future (as seems likely... 😅).

**Comment by ppwadhwa** [comment]
@jni 
@napari/core-devs 

oops, now I realize I have let this one drop.  :(. Are we still interested in this?  I will spend some time to resolve conflicts and looking at the tests.  I was looking for this in preferences, because I remembered working on this, and then realized this never got merged.  So sorry, y'all.

**Comment by kephale** [comment]
@ppwadhwa yes, this is getting more and more important because we have some extra Dask stuff on the horizon. Thank you so much!

**Comment by jni** [comment]
@ppwadhwa it is I who is sorry that I didn't re-review for so long! If you could push this over the line in the coming days and ping me in `#reviews` on Zulip when done, I would be most grateful! 🙏

**Comment by goanpeca** [comment]
@Czaki do you have some ideas on how we can fix the problematic test?

It seems the magicgui widgets are being collected prematurely by the GC. See https://github.com/napari/napari/actions/runs/3678867738/jobs/6222639125#step:8:408

Thanks!

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_dask_settings.py]:
```suggestion
from magicgui import magic_factory
```

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_dask_settings.py]:
```suggestion
@magic_factory(auto_call=True, layout='horizontal', cache={'min': 0, 'max': 20})
```

**Review by Czaki — COMMENTED** [review]
We should not use `magicgui` decorator, but `magic_factory` that will create separate widgets, not reuse existing ones.

**Comment by ppwadhwa** [comment]
This PR is so close, but this last test has me stumped.  Has anyone seen this error before?  

```
E   SystemError: <class 'Shiboken.ObjectType'> returned a result with an exception set
```

@Czaki @goanpeca @kephale @jni

**Comment by Czaki** [comment]
I can confirm that I'm able to reproduce it locally on ubuntu, even using napari (just open window, and try to open preferences dialog).

**Comment by Czaki** [comment]
@ppwadhwa Is there any reason why we would like to have an option of implementing settings via magicgui decorators? 

To debug this, I need most probably reimplement this using pure Qt, so is there a reason why we need to use magicgui widgets (widgets factory) instead of providing just Qt widgets, or some form of JSON schema?

**Comment by ppwadhwa** [comment]
@Czaki I had originally coded this using pure QT (which was probably much more complicated than it needed to be with the way I did it), then it was suggested to use magicgui.

**Comment by jni** [comment]
@Czaki yeah, sorry that this review got a bit long but @ppwadhwa's summary gets to the gist: ultimately, I think that we don't want to manually write a whole bunch of widget and UI chrome for each setting. It's just a lot of boilerplate that imho we should try to avoid in napari. magicgui seems like the right tool to achieve setting <> UI mappings.

**Comment by Czaki** [comment]
> @Czaki yeah, sorry that this review got a bit long but @ppwadhwa's summary gets to the gist: ultimately, I think that we don't want to manually write a whole bunch of widget and UI chrome for each setting. It's just a lot of boilerplate that imho we should try to avoid in napari. magicgui seems like the right tool to achieve setting <> UI mappings.

So when I try to replace magic GUI with a custom widget I just found that there is boilerplate code connected with this setting in `qt_jsonschema_form` module (it is strange that we need to modify vendored module - vendored should be not changed or moved outside `_vendored`). 

At this moment I could reproduce it locally but do not have any idea how to debug this (except build pyside2 from source in debug mode, read c++ code to understand it, set proper breakpoints and execute everything under gdb). 

As longer I think about this, We should generate (using magicgui) settings from JSON-schema or pydantic BaseModel. Rather than glue directly to widgets generated from magicgui. But at least we should have some MguiSettingsWidget, one for all widgets. 

PS. If someone would have any idea how to debug the current problem I will be happy to help.

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py]:
```suggestion
```

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py]:
```suggestion
    valueChanged = Signal(dict)
    
    def get_mgui(self):
        """Get magic gui widget."""
        raise NotImplementedError()


    def __init__(self, description=None):
```

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py]:
```suggestion
        self._widget = self.get_mgui()
```

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_dask_settings.py]:
```suggestion
    def get_mgui(self):
        return dask_settings()
```

**Review by Czaki — CHANGES_REQUESTED** [review]
After these changes, the test pass locally. 

It looks like `MAGIC_GUI` was treated as function ad pyside fails performing some checks.

**Comment by ppwadhwa** [comment]
@Czaki thank you!!! you are awesome!  I committed all the changes and see it looks fine on my computer.  Fingers crossed all the tests pass here.  :)

**Comment by Czaki** [comment]
Small reminder. In the Files Changed tab, you could add all suggestions to the batch and create one commit.

**Comment by Czaki** [comment]
It looks like it is working. I need to cancel a few jobs to not wait too long. 

I have opened #5453 to not need to do such activities in the future.

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py:45]:
```suggestion
            return getattr(self._widget.native, attribute)
```

**Comment by ppwadhwa** [comment]
I just rebased.  Hopefully, this is good to go for now to merge?  

@Czaki @jni @kephale @goanpeca

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py]: Is this required? Why we cannot jut set tooltip? This is a private API of magicgui and could be broken in any release.

File: napari/settings/_application.py
Comment (Czaki) [inline napari/settings/_application.py:192]: Why did we need to have this as one field, not two?

**Review by Czaki — COMMENTED** [review]
A few questions/remarks.

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (ppwadhwa) [reply napari/_qt/widgets/magic_gui_base_widget.py]: It turns out that this was no longer working to set the tooltip.  And I'm not able to figure out how to set a tooltip otherwise, but with the changes, I made, it works.

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py]: Still, we use here magicgui private API. It is a really bad idea. If it is not possible to be done with public API then better not do this or ask @tlambert03 for a hint.

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py
Comment (Czaki) [inline napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py]: why it is here, not in `MagicDaskSettingsWidget`?

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py
Comment (Czaki) [inline napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py]: again private api (our api) instead of proper method/property in `BaseMagicSetting`

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py]: Why is this not QWidget? 

```suggestion
class BaseMagicSetting(QWidget):
    """Base class that helps use Magic GdUI widgets in the json schema widget
    buider.
    """

    valueChanged = Signal(dict)

    def get_mgui(self):
        """Get magic gui widget."""
        raise NotImplementedError()

    def __init__(self, description=None):
        super().__init__()
        self._widget = self.get_mgui()
        self._description = description
        
        self.setLayout(QHBoxLayout())
        self.layout().addWidget(self._widget.native)

        self._widget.changed.connect(self._call_magic_gui)

    def _call_magic_gui(self, value: dict = None):
        if self.value is None:
            value = {}
        self.valueChanged.emit(value)
```

**Review by Czaki — COMMENTED** [review]
At this moment, for me, this PR is somewhere between one of two optimal states.

It should either be a really simple pr (for example add two fields to the application settings) or should prepare napari on any magicgui base settings. 

It looks closer to the second one but still has dask specific code in the the`qt_json_builder` instead of the generic magicgui one.

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (tlambert03) [reply napari/_qt/widgets/magic_gui_base_widget.py]: Get_mgui() is abstract here and doesn't have a type annotation, so I can't tell what `self._widget` actually is.  Not all magicgui widgets have a `_list` attribute, so that's also bad.

But, if it's a Container, just iterate over self._widget directly.  (But I would definitely clarify here what get_mgui is and isn't allowed to return)

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_dask_settings.py:18]: @tlambert03 The only implementation of get_mgui is here. MAy you suggest what type of annotation should be used?

Comment (tlambert03) [reply napari/_qt/widgets/magic_gui_dask_settings.py:18]: Dask_settings appears to be a FunctionGui (which is a Container, which is why this works), but you'd have to ask @ppwadhwa whether the design of this abstract class demands that or not

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py
Comment (ppwadhwa) [reply napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py]: I am not sure how to answer this, except that I followed a similar pattern to every other widget being used in the preferences dialog.  Should it be moved?  @goanpeca do you have any thoughts on this?

Comment (Czaki) [reply napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py]: In `qt_jsonschema_form` the code should be generic, not specific.

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (ppwadhwa) [reply napari/_qt/widgets/magic_gui_dask_settings.py:18]: @tlambert03 @Czaki I am not sure.  In the end, I just want a widget with a checkbox and spinbox in a horizontal layout to be added to the preferences dialog.  I am trying to use MagicGui to do this, and at the same time trying to fit it all into the json schema builder so it can work the same as every other widget in the preferences dialog.  Should I be doing things differently?  I would appreciate any suggestions on how to move forward.

**Comment by ppwadhwa** [comment]
@Czaki I really appreciate your comments.  I am going to change things to be more generic in the form.py file as you said.  I will also keep trying to figure out how to use the public api for setting the tooltip.

File: napari/_qt/widgets/magic_gui_dask_settings.py
Comment (tlambert03) [reply napari/_qt/widgets/magic_gui_dask_settings.py:18]: This widget is fine.  The issue is over in the base widget, where you've created what appears to be an abstract class for "any generic widget that magicgui might provide"... but then you've used a private attribute specifically from magicgui Containers. So, while this works fine for your one dask widget, someone might reasonably assume that they can use your base class for _any_ magicgui widget (even those which are not containers) and then your setTooltip method will raise an exception

Comment (ppwadhwa) [reply napari/_qt/widgets/magic_gui_dask_settings.py:18]: @tlambert03 thank you for the explanation.  I understand the problem now.

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (ppwadhwa) [reply napari/_qt/widgets/magic_gui_base_widget.py]: I made these changes.  Thank you.

**Comment by Czaki** [comment]
> @Czaki I really appreciate your comments. I am going to change things to be more generic in the form.py file as you said. I will also keep trying to figure out how to use the public api for setting the tooltip.

If I'm too laconic please ping me (either here or on zulip). It is sometimes hard to estimate knowledge level and maybe I'm assuming that something is obvious if it is not true in your case.

File: napari/settings/_application.py
Comment (ppwadhwa) [reply napari/settings/_application.py:192]: One field will mean one widget.  If it is 2, it will be 2 lines on the preferences dialog.

**Comment by ppwadhwa** [comment]
@Czaki I made some changes.  What do you think?

File: napari/_qt/qt_viewer.py
Comment (Czaki) [inline napari/_qt/qt_viewer.py:304]:
```suggestion
        settings.application.events.dask.connect(self._update_dask_settings)
```

File: napari/_qt/widgets/magic_gui_base_widget.py
Comment (Czaki) [inline napari/_qt/widgets/magic_gui_base_widget.py:32]: dict is not callable and one line above, you assign the default value as dict.

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/form.py
Comment (Czaki) [inline napari/_vendor/qt_json_builder/qt_jsonschema_form/form.py:34]: Where the name `dask` is set? It is a field name?

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py
Comment (Czaki) [inline napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py:611]: These two may live in BaseMagicSetting As they will be shared by any magicgui based settings field?

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py
Comment (Czaki) [inline napari/_vendor/qt_json_builder/qt_jsonschema_form/widgets.py:615]: maybe it could be implemented directly in MagicDaskSettingsWidge?

**Review by Czaki — COMMENTED** [review]
I need to test a few inner mechanisms to understand if something could be improved.

File: napari/_vendor/qt_json_builder/qt_jsonschema_form/form.py
Comment (ppwadhwa) [reply napari/_vendor/qt_json_builder/qt_jsonschema_form/form.py:34]: In `preferences_dialog.py`, there is a reference dictionary where you specify certain widgets to be used when the fields from certain settings are to be built.  The value used here in `form.py` is set there, but it matches a field in settings.
ui_schema = {
        "call_order": {"ui:widget": "plugins"},
        "highlight_thickness": {"ui:widget": "highlight"},
        "shortcuts": {"ui:widget": "shortcuts"},
        "extension2reader": {"ui:widget": "extension2reader"},
        "dask": {"ui:widget": "dask"},
    }

**Comment by jni** [comment]
Closed in favour of #5490

**Pull request closed without merging**
